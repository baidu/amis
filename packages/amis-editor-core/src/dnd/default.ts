/**
 * 默认的拖拽模式实现。
 */
import {animation} from 'amis';
import findIndex from 'lodash/findIndex';
import {EditorDNDManager} from '.';
import {renderThumbToGhost} from '../component/factory';
import {EditorNodeType} from '../store/node';
import {translateSchema} from '../util';
import {DNDModeInterface} from './interface';

export class DefaultDNDMode implements DNDModeInterface {
  readonly dndContainer: HTMLElement; // 记录当前拖拽区域
  dropBeforeId?: string;
  constructor(readonly dnd: EditorDNDManager, readonly region: EditorNodeType) {
    // 初始化时，默认将元素所在区域设置为当前拖拽区域
    this.dndContainer = this.dnd.store
      .getDoc()
      .querySelector(
        `[data-region="${region.region}"][data-region-host="${region.id}"]`
      ) as HTMLElement;
  }

  /**
   * 记录上次交换时的鼠标位置。
   */
  exchangeX: number = 0;
  exchangeY: number = 0;

  /**
   * 首次拖入，把 ghost 插入进来。让用户有个直观感受。
   * @param e
   * @param ghost
   */
  enter(e: DragEvent, ghost: HTMLElement) {
    const dragEl = this.dnd.dragElement;
    // 如果就在移动的元素所在的区域，那么把 ghost 放到原来的位置。
    if (dragEl && dragEl.closest('[data-region]') === this.dndContainer) {
      const list: Array<any> = Array.isArray(this.region.schema)
        ? this.region.schema
        : [];
      const child = this.getChild(this.dndContainer, dragEl);
      const id = dragEl.getAttribute('data-editor-id')!;
      const idx = findIndex(list, (item: any) => item.$$id === id);
      if (~idx && list[idx + 1]) {
        this.dropBeforeId = list[idx + 1].$$id;
      }
      this.dndContainer.insertBefore(ghost, child);

      let innerHTML = dragEl.outerHTML
        .replace('ae-is-draging', '')
        // .replace(/\bdata\-editor\-id=('|").+?\1/g, '');
        // 上面那个会让 fis 编译的时候把后续的 require 语句都给忽略掉了。
        .replace(/\bdata\-editor\-id=(?:'.+?'|".+?")/g, '');
      // bca-disable-next-line
      ghost.innerHTML = innerHTML;
    } else {
      const manager = this.dnd.manager;
      const store = manager.store;
      renderThumbToGhost(
        ghost,
        this.region,
        translateSchema(store.dragSchema),
        manager
      );
      this.dndContainer.appendChild(ghost);
    }
  }

  /**
   * 拖出去了，就移除 ghost
   * @param e
   * @param ghost
   */
  leave(e: DragEvent, ghost: HTMLElement) {
    this.dndContainer.removeChild(ghost);
  }

  over(e: DragEvent, ghost: HTMLElement) {
    const target = this.getTarget(e);
    const wrapper = this.dndContainer;
    const elemSchema = this.region.schema;
    const list: Array<any> = Array.isArray(elemSchema) ? elemSchema : [];
    const dx = e.clientX - this.exchangeX;
    const dy = e.clientY - this.exchangeY;
    const vertical = Math.abs(dy) > Math.abs(dx);

    if (target && !animation.animating) {
      const targetId = target.getAttribute('data-editor-id')!;
      const targetChild = this.getChild(wrapper, target!);
      const idx = findIndex(list, (item: any) => item.$$id === targetId);

      const originIdx = Array.prototype.indexOf.call(wrapper.children, ghost);
      const targetIdx = Array.prototype.indexOf.call(
        wrapper.children,
        targetChild
      );

      if (
        ~originIdx &&
        originIdx > targetIdx &&
        (!this.exchangeY || dy < 0 || dx < 0)
      ) {
        // 原来在后面，移动到前面

        this.exchangeX = e.clientX;
        this.exchangeY = e.clientY;
        this.dropBeforeId = list[idx]?.$$id;

        if (originIdx !== targetIdx - 1) {
          animation.capture(wrapper);
          wrapper.insertBefore(ghost, targetChild);
          animation.animateAll();
        }
      } else if (
        ~originIdx &&
        originIdx < targetIdx &&
        (!this.exchangeY || dy > 0 || dx > 0)
      ) {
        // 原来在前面，移动到后面

        this.exchangeX = e.clientX;
        this.exchangeY = e.clientY;
        if (list[idx + 1]) {
          this.dropBeforeId = list[idx + 1]?.$$id;
        } else {
          delete this.dropBeforeId;
        }

        if (originIdx !== targetIdx + 1) {
          animation.capture(wrapper);
          wrapper.insertBefore(ghost, targetChild.nextSibling);
          animation.animateAll();
        }
      }
    }

    if (ghost.parentNode !== wrapper) {
      delete this.dropBeforeId;
      animation.capture(wrapper);
      wrapper.appendChild(ghost);
      animation.animateAll();
    }
  }

  /**
   * 返回个相对位置，如果没有数据会插入到结尾。
   */
  getDropBeforeId() {
    return this.dropBeforeId;
  }

  /**
   * 获取当时拖动到了哪个节点上面。
   */
  getTarget(e: DragEvent) {
    let target = (e.target as HTMLElement).closest(
      '[data-editor-id]'
    ) as HTMLElement;

    while (target) {
      const region = target.parentElement?.closest('[data-region]');

      if (region === this.dndContainer) {
        const renderer = target.getAttribute('data-renderer');
        if (renderer === 'grid') {
          // grid 组件中的分栏的子栏也可以拖入 选中分栏组件同级组件拖动时有问题 兼容一下
          return target.parentElement;
        } else {
          return target;
        }
      }

      target =
        (target.parentElement?.closest('[data-editor-id]') as HTMLElement) ||
        null;
    }

    return null;
  }

  /**
   * 获取区域的直接孩子，因为有时候会在孩子的孩子里面。
   * 但是插入 ghost 的相对位置，insertBefore 只能是当前孩子。
   * @param dom
   * @param descend
   */
  getChild(dom: HTMLElement, descend: HTMLElement) {
    let child = descend;

    while (child) {
      if (child.parentElement === dom) {
        break;
      }

      child = child.parentElement!;
    }

    return child;
  }

  /**
   * 销毁
   */
  dispose() {
    delete this.dropBeforeId;
  }
}
