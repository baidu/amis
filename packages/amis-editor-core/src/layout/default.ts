import {LayoutInterface} from './interface';

export default class DefaultLayout implements LayoutInterface {}
