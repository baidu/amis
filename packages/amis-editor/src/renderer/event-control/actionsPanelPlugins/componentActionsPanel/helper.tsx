import {JSONGetById, EditorManager} from 'amis-editor-core';
import {DataSchema} from 'amis-core';
import CmptActionSelect from '../../comp-action-select';

// 下拉展示可赋值属性范围
export const SELECT_PROPS_CONTAINER = ['form'];

export const renderCmptActionSelect = (
  componentLabel: string,
  required: boolean,
  onChange?: (value: string, oldVal: any, data: any, form: any) => void,
  hideAutoFill?: boolean,
  manager?: EditorManager
) => {
  return [
    ...renderCmptSelect(
      componentLabel || '选择组件',
      true,
      async (value: string, oldVal: any, data: any, form: any) => {
        // 获取组件上下文
        if (form.data.__nodeId) {
          if (form.data.actionType === 'setValue') {
            // todo:这里会闪一下，需要从amis查下问题
            form.setValueByName('args.value', []);
            form.setValueByName('args.__comboType', undefined);
            form.setValueByName('args.__valueInput', undefined);
            form.setValueByName('args.__containerType', undefined);

            if (SELECT_PROPS_CONTAINER.includes(form.data.__rendererName)) {
              const contextSchema: any = await form.data.getContextSchemas?.(
                form.data.__nodeId,
                true
              );

              const dataSchema = new DataSchema(contextSchema || []);
              const variables = dataSchema?.getDataPropsAsOptions() || [];
              form.setValueByName(
                '__setValueDs',
                variables.filter(item => item.value !== '$$id')
              );
            } else {
              form.setValueByName('__setValueDs', []);
            }
          }
        }
        form.setValueByName('groupType', '');
        onChange?.(value, oldVal, data, form);
      },
      hideAutoFill
    ),
    {
      type: 'input-text',
      name: '__cmptId',
      mode: 'horizontal',
      size: 'lg',
      required: true,
      label: '组件id',
      visibleOn:
        'this.componentId === "customCmptId" && this.actionType === "component"',
      onChange: async (value: string, oldVal: any, data: any, form: any) => {
        let schema = JSONGetById(manager!.store.schema, value, 'id');
        if (schema) {
          form.setValues({
            __rendererName: schema.type
          });
        } else {
          form.setValues({
            __rendererName: ''
          });
        }
      }
    },
    {
      asFormItem: true,
      label: '组件动作',
      name: 'groupType',
      mode: 'horizontal',
      required: true,
      visibleOn: 'this.actionType === "component"',
      component: CmptActionSelect,
      description: '${__cmptActionDesc}'
    }
  ];
};

export const renderCmptSelect = (
  componentLabel: string,
  required: boolean,
  onChange?: (value: string, oldVal: any, data: any, form: any) => void,
  hideAutoFill?: boolean
) => [
  {
    type: 'tree-select',
    name: 'componentId',
    label: componentLabel || '选择组件',
    showIcon: false,
    searchable: true,
    required,
    selfDisabledAffectChildren: false,
    size: 'lg',
    source: '${__cmptTreeSource}',
    mode: 'horizontal',
    autoFill: {
      __isScopeContainer: '${isScopeContainer}',
      ...(hideAutoFill
        ? {}
        : {
            __rendererLabel: '${label}',
            __rendererName: '${type}',
            __nodeId: '${id}',
            __nodeSchema: '${schema}'
          })
    },
    onChange: async (value: string, oldVal: any, data: any, form: any) => {
      onChange?.(value, oldVal, data, form);
    }
  }
];

export const renderCmptIdInput = (
  onChange?: (value: string, oldVal: any, data: any, form: any) => void
) => {
  return {
    type: 'input-text',
    name: '__cmptId',
    mode: 'horizontal',
    size: 'lg',
    required: true,
    label: '组件id',
    visibleOn: 'this.componentId === "customCmptId"',
    onChange: async (value: string, oldVal: any, data: any, form: any) => {
      onChange?.(value, oldVal, data, form);
    }
  };
};
