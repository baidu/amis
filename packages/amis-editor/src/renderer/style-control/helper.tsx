import {getSchemaTpl} from 'amis-editor-core';

export const inputStateTpl = (className: string, path: string = '') => {
  return [
    {
      type: 'select',
      name: 'editorState',
      label: '状态',
      selectFirst: true,
      options: [
        {
          label: '常规',
          value: 'default'
        },
        {
          label: '悬浮',
          value: 'hover'
        },
        {
          label: '点击',
          value: 'active'
        }
      ]
    },
    ...inputStateFunc(
      "${editorState == 'default' || !editorState}",
      'default',
      className,
      path
    ),
    ...inputStateFunc("${editorState == 'hover'}", 'hover', className, path),
    ...inputStateFunc("${editorState == 'active'}", 'active', className, path)
  ];
};

export const inputStateFunc = (
  visibleOn: string,
  state: string,
  className: string,
  path: string,
  options: any = []
) => {
  return [
    getSchemaTpl('theme:font', {
      label: '文字',
      name: `${className}.font:${state}`,
      visibleOn: visibleOn,
      editorThemePath: `${path}.${state}.body.font`
    }),
    getSchemaTpl('theme:colorPicker', {
      label: '背景',
      name: `${className}.background:${state}`,
      labelMode: 'input',
      needGradient: true,
      needImage: true,
      visibleOn: visibleOn,
      editorThemePath: `${path}.${state}.body.bg-color`
    }),
    getSchemaTpl('theme:border', {
      name: `${className}.border:${state}`,
      visibleOn: visibleOn,
      editorThemePath: `${path}.${state}.body.border`
    }),
    getSchemaTpl('theme:paddingAndMargin', {
      name: `${className}.padding-and-margin:${state}`,
      visibleOn: visibleOn,
      editorThemePath: `${path}.${state}.body.padding-and-margin`
    }),
    getSchemaTpl('theme:radius', {
      name: `${className}.radius:${state}`,
      visibleOn: visibleOn,
      editorThemePath: `${path}.${state}.body.border`
    }),
    ...options
  ];
};
