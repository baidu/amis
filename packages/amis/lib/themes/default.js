"use strict";
/**
 * @file 这个文件只作为兼容，后续没有 default 了，default 就是 cxd
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
(0, tslib_1.__exportStar)(require("./cxd"), exports);
//# sourceMappingURL=./themes/default.js.map
