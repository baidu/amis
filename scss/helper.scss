@import './functions';
@import './variables';
@import './mixins';

// background
@import './helper/background/background-color';

@import './helper/border/border-color';
@import './helper/border/border-radius';
@import './helper/border/border-width';
@import './helper/border/border-style';

@import './helper/layout/box-sizing';
@import './helper/layout/clear';
@import './helper/layout/display';
@import './helper/layout/float';
@import './helper/layout/overflow';
@import './helper/layout/position';
@import './helper/layout/top-right-bottom-left';
@import './helper/layout/visibility';
@import './helper/layout/z-index';

@import './helper/flex/direction';
@import './helper/flex/flex';
@import './helper/flex/grow';
@import './helper/flex/order';
@import './helper/flex/shrink';
@import './helper/flex/wrap';

@import './helper/grid/auto-columns';
@import './helper/grid/auto-flow';
@import './helper/grid/auto-rows';
@import './helper/grid/column-start-end';
@import './helper/grid/columns';
@import './helper/grid/gap';
@import './helper/grid/row-start-end';
@import './helper/grid/rows';

@import './helper/box-alignment/align-content';
@import './helper/box-alignment/align-items';
@import './helper/box-alignment/align-self';
@import './helper/box-alignment/justify-content';
@import './helper/box-alignment/justify-items';
@import './helper/box-alignment/justify-self';
@import './helper/box-alignment/place-content';
@import './helper/box-alignment/place-items';
@import './helper/box-alignment/place-self';

@import './helper/effect/box-shadow';

@import './helper/spacing/margin';
@import './helper/spacing/padding';
@import './helper/spacing/space-between';

@import './helper/sizing/width';
@import './helper/sizing/height';

@import './helper/typography/font-size';
@import './helper/typography/font-style';
@import './helper/typography/font-weight';
@import './helper/typography/line-height';
@import './helper/typography/list-style-type';
@import './helper/typography/text-align';
@import './helper/typography/text-color';
@import './helper/typography/text-decoration';
@import './helper/typography/text-transform';
@import './helper/typography/whitespace';
@import './helper/typography/word-break';
