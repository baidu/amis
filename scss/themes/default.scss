// 默认主题
@import './default-variables';

@import './common';
