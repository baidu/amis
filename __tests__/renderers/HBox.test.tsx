import React = require('react');
import {render} from 'react-testing-library';
import '../../src/themes/default';
import {
    render as amisRender
} from '../../src/index';
import { makeEnv } from '../helper';

test('Renderer:hbox', async () => {
    const {
        container
    } = render(amisRender({
        type: 'hbox',
        columns: [
            {
                type: 'tpl',
                tpl: 'w-xs',
                className: 'bg-info',
                inline: false,
                columnClassName: 'w-xs'
            },
            {
                type: 'tpl',
                tpl: 'w-sm',
                className: 'bg-info lter',
                inline: false,
                columnClassName: 'w-sm'
            },
            {
                type: 'tpl',
                tpl: 'w',
                className: 'bg-info dk',
                inline: false,
                columnClassName: 'w'
            },
            {
                type: 'tpl',
                tpl: '平均分配',
                className: 'bg-success',
                inline: false
            },
            {
                type: 'tpl',
                tpl: '平均分配',
                className: 'bg-primary',
                inline: false
            }
        ]
    }, {}, makeEnv({
    })));

    expect(container).toMatchSnapshot();
});