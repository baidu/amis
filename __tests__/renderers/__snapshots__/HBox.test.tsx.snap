// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer:hbox 1`] = `
<div>
  <div
    class="a-Hbox"
  >
    <div
      class="a-Hbox-col w-xs"
    >
      <div
        class="a-TplField bg-info"
      >
        <span>
          w-xs
        </span>
      </div>
    </div>
    <div
      class="a-Hbox-col w-sm"
    >
      <div
        class="a-TplField bg-info lter"
      >
        <span>
          w-sm
        </span>
      </div>
    </div>
    <div
      class="a-Hbox-col w"
    >
      <div
        class="a-TplField bg-info dk"
      >
        <span>
          w
        </span>
      </div>
    </div>
    <div
      class="a-Hbox-col"
    >
      <div
        class="a-TplField bg-success"
      >
        <span>
          平均分配
        </span>
      </div>
    </div>
    <div
      class="a-Hbox-col"
    >
      <div
        class="a-TplField bg-primary"
      >
        <span>
          平均分配
        </span>
      </div>
    </div>
  </div>
</div>
`;
